pyiron_base/database/performance.py

```diff
@@ -0,0 +1,170 @@
+import pandas as pd
+from sqlalchemy import (
+    create_engine,
+    select,
+    distinct,
+    MetaData,
+    Table,
+    func,
+    or_,
+    false,
+)
+from pyiron_base.settings.generic import Settings
+
+
+__author__ = "Muhammad Hassani"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Muhammad Hassani"
+__email__ = "hassani@mpie.de"
+
+
+class DatabaseStatistics:
+    def __init__(self):
+        s = Settings()
+        self._connection_string = s._configuration['sql_connection_string']
+        self._job_table = s._configuration['sql_view_table_name']
+        if "postgresql" not in self._connection_string:
+            raise RuntimeError(
+                """
+                The detabase statistics is only available for a Postgresql database
+                """
+            )
+        self._table = s._configuration['sql_table_name']
+        self._engine = create_engine(self._connection_string)
+        self._performance_dict = {}
+        self.total_index_size = 0
+        self._metadata = MetaData()
+        self._stat_view = Table('pg_stat_activity', self._metadata, autoload_with=self._engine)
+        self._locks_view = Table('pg_locks', self._metadata, autoload_with=self._engine)
+
+    def _num_conn(self, conn):
+        """
+        return the number of connections
+        """
+        stmt = select(func.count()).select_from(self._stat_view)
+        result = conn.execute(stmt)
+        self._performance_dict['total num. connection'] = result.fetchone()[0]
```

Comment:
I find this code hard to read. From my perspective it would be easier to return the entry rather than writing to the dictionary. This also simplifies debugging later.

Reply:
Do you mean the whole code or this function in particular? There are multiple functions which modifies the `self._performance_dict`. Some of them, like [here](https://github.com/pyiron/pyiron_base/blob/b1c02b722ce05a4dde3225e5d7ef73f93799d9dd/pyiron_base/database/performance.py#L60) and [here](https://github.com/pyiron/pyiron_base/blob/b1c02b722ce05a4dde3225e5d7ef73f93799d9dd/pyiron_base/database/performance.py#L133) would be more complex to be returned an entry. But definitely can be done.

Reply:
Yes, I would like each function to return a sub dictionary and then use `dict.update()` to merge them in the end. https://python-reference.readthedocs.io/en/latest/docs/dict/update.html This way we can debug the individual functions, look at the output they return. Following the zen of python: Explicit is better than implicit. https://www.python.org/dev/peps/pep-0020/

Reply:
Thanks for the links. Now all methods return a dictionary which is used at the end to update `self._performance_dict`.